cms/datavis/blocks/utils.py

```diff
@@ -0,0 +1,25 @@
+from django.forms import TextInput
+from wagtail import blocks
+
+
+class TextInputCharBlock(blocks.IntegerBlock):
+    """A text input widget that only allows numeric input.
+
+    See https://design-system.service.gov.uk/components/text-input/#numbers
+    """
+
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+        self.field.widget = TextInput(attrs={"inputmode": "numeric", "pattern": "[0-9]*"})
+
+
+class TextInputFloatBlock(blocks.FloatBlock):
+    """A text input widget that only allows decimal input.
+
+    See https://design-system.service.gov.uk/components/text-input/#numbers
+    """
+
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+        # NB inputmode is not "decimal" as per the guidance linked in the docstring above
+        self.field.widget = TextInput(attrs={"inputmode": "text", "pattern": "[0-9]*\\.?[0-9]*"})
```

Comment:
This might be nit-picky but the pattern allows to input just `.` (also numbers without leading or trailing zeroes).

How about something like this?

```suggestion
        self.field.widget = TextInput(attrs={"inputmode": "text", "pattern": "\\d*(\\.\\d+)?"})
```

(We could also use a raw string to avoid the double slashes: `r"\d*(\.\d+)?"`)

Reply:
Thanks for calling me out on this. The integer input pattern was verbatim from the GDS blog post, but I think the float one was suggested by autocomplete in my code editor, and I admit I did not check it.

I spent some time this morning thinking about edge cases, and doing TDD, and went down a rabbit hole of whether it was right to permit inputs like `0123` (integer with leading zeroes), `-0` (negative zero), `.5` (shorthand decimal), or `3.`, i.e. things which are legitimate but not _well-formed_.

Then I saw that the GDS docs examples at https://design-system.service.gov.uk/components/text-input/#numbers and https://design-system.service.gov.uk/components/text-input/#asking-for-decimal-numbers don't have patterns at all, and remembered that GDS advises to disable browser validation with `novalidate`. Wagtail already does this, so I think we can remove these regex patterns, and my tests can burn.

For posterity, I settled on:

```python
INTEGER_PATTERN = r"^-?([1-9][0-9]*|0)$"
FLOAT_PATTERN = r"^-?([1-9][0-9]*|0)(\.[0-9]+)?$"
```

Reply:
Novalidate docs ref: https://design-system.service.gov.uk/patterns/validation/#turn-off-html5-validation